bazel/pch.bzl

```diff
@@ -0,0 +1,114 @@
+load(
+    "@bazel_tools//tools/build_defs/cc:action_names.bzl",
+    "CPP_COMPILE_ACTION_NAME",
+)
+
+def _pch(ctx):
+    deps_cc_info = cc_common.merge_cc_infos(
+        cc_infos = [dep[CcInfo] for dep in ctx.attr.deps],
+    )
+
+    if not ctx.attr.enabled:
+        return [deps_cc_info]
+
+    cc_toolchain = ctx.attr._cc_toolchain[cc_common.CcToolchainInfo]
+    feature_configuration = cc_common.configure_features(
+        ctx = ctx,
+        cc_toolchain = cc_toolchain,
+        requested_features = ctx.features,
+        unsupported_features = ctx.disabled_features,
+    )
+
+    cc_compiler_path = cc_common.get_tool_for_action(
+        feature_configuration = feature_configuration,
+        action_name = CPP_COMPILE_ACTION_NAME,
+    )
+
+    if "clang" not in cc_compiler_path:
+        fail("error: attempting to use clang PCH without clang: {}".format(cc_compiler_path))
+
+    generated_header_file = ctx.actions.declare_file(ctx.label.name + ".h")
+    ctx.actions.write(
+        generated_header_file,
+        "\n".join(["#include \"{}\"".format(include) for include in ctx.attr.includes]) + "\n",
+    )
+
+    # TODO: -fno-pch-timestamp / invalidation in that case doesn't work
+    pch_flags = ["-x", "c++-header"]
```

Comment:
Do you also want to use `-fpch-instantiate-templates` or `-fpch-codegen` and `-fpch-debuginfo` ?

Reply:
Oh, I see. `-fpch-codegen` and `-fpch-debuginfo` actually require compiling and then linking the resulting object file. We could add it in the next ireation.
I think the `-fpch-instantiate-templates` can be added without the need to compile and link the precompiled header file.

Reply:
Right now `-fpch-instantiate-templates` isn't part of the newest clang released with Xcode on macOS, I'm not sure how we want to handle things like that but I didn't consider adding any options in that situation since I've been building this on macOS